### PR TITLE
ENH: Resample volume to resolution

### PR DESCRIPTION
**Pull request by AntoineTheb**
`scil_resample_volume` now handles resampling to arbitrary resolutions. Will be useful for future multi-resolution tracking.
Also fixes a bug with the offset of resampled volumes. "Old" behavior is still available through the `zoom` argument.

Test data: https://drive.google.com/file/d/1k0vqk4tp5y0q3SRGOtU2esY2a8h0zGT8/view?usp=sharing

To test:
`$ scil_resample_volume.py Simulated_FiberCup_fodf_6.nii.gz Simulated_FiberCup_fodf_6_32_32_3.nii.gz --resolution 32 32 3 -f`
Downsamples the fibercup from 64x64x3 to 32x32x3

@arnaudbore @frheault @GuillaumeTh @mdesco

### CODE REVIEW

File: scilpy/image/resample_volume.py [hunk 1]
```diff
@@ -46,23 +46,32 @@ def resample_volume(img, ref=None, res=None, iso_min=False, interp='lin',
         Resampled image.
     """
     data = img.get_fdata(dtype=np.float32)
```
Comment (AntoineTheb) [inline scilpy/image/resample_volume.py:48]: @arnaudbore Data is always loaded as float32, which breaks masks (can't be loaded with `load_data_as_mask` anymore) if you try to downsample one. What's the solution to this ?

Comment (arnaudbore) [reply scilpy/image/resample_volume.py:48]: Do you think the resampling algorithm could be different with masks apart from the loading part ?

Comment (arnaudbore) [reply scilpy/image/resample_volume.py:48]: Add zoom in Parameters

Comment (arnaudbore) [reply scilpy/image/resample_volume.py:48]: You could add a boolean to say if it's a mask or not.

Comment (arnaudbore) [reply scilpy/image/resample_volume.py:48]: Or find the data type and load accordingly

**Review by arnaudbore — COMMENTED** [review]
Few comments

File: scilpy/image/resample_volume.py (hunk 1 above)
Comment (AntoineTheb) [reply scilpy/image/resample_volume.py:48]: Found how to load data as-is here: https://nipy.org/nibabel/reference/nibabel.dataobj_images.html?highlight=get_fdata#nibabel.dataobj_images.DataobjImage.get_data

Will update.

File: scilpy/image/resample_volume.py [hunk 2]
```diff
@@ -46,23 +46,32 @@ def resample_volume(img, ref=None, res=None, iso_min=False, interp='lin',
...
+            res = res * 3
+        new_zooms = tuple((o / r) * z for o, r,
+                          z in zip(original_res, res, original_zooms))
+
     elif iso_min:
-        min_zoom = min(original_zooms)
-        new_zooms = (min_zoom, min_zoom, min_zoom)
+        if zoom:
+            min_zoom = min(original_zooms)
+            new_zooms = (min_zoom, min_zoom, min_zoom)
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scilpy/image/resample_volume.py:72]: I messed up the logic there, if `iso_min` and `zoom` should raise an error. Will fix.

[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scilpy/image/resample_volume.py]: Fixed in 7904f52

File: scilpy/image/resample_volume.py (hunk 1 above)
Comment (AntoineTheb) [reply scilpy/image/resample_volume.py]: Donesies in 7904f52

File: scilpy/image/reslice.py [hunk 3]
```diff
@@ -16,7 +16,7 @@ def _affine_transform(kwargs):
 
 
 def reslice(data, affine, zooms, new_zooms, order=1, mode='constant', cval=0,
-            num_processes=1):
+            offset=-0.5, num_processes=1):
```
Comment (arnaudbore) [inline scilpy/image/reslice.py:19]: add offset in the doc

File: scilpy/image/resample_volume.py [hunk 4]
```diff
@@ -12,8 +12,8 @@ def _interp_code_to_order(interp_code):
     return orders[interp_code]
 
 
-def resample_volume(img, ref=None, res=None, iso_min=False, interp='lin',
-                    enforce_dimensions=False):
+def resample_volume(img, ref=None, res=None, iso_min=False, zoom=None,
```
Comment (arnaudbore) [inline scilpy/image/resample_volume.py:15]: add zoom in the doc

**Review by arnaudbore — APPROVED** [review]
Few doc missing but LGTM.

File: scilpy/image/reslice.py (hunk 3 above)
Comment (AntoineTheb) [reply scilpy/image/reslice.py:19]: Fixed in 0a0c708

File: scilpy/image/resample_volume.py (hunk 4 above)
Comment (AntoineTheb) [reply scilpy/image/resample_volume.py:15]: Fixed in 0a0c708

**Comment by arnaudbore** [comment]
I just talked to @GuillaumeTh and the vox_offset is a pretty tricky thing to play with. [Some documentation](https://nifti.nimh.nih.gov/nifti-1/documentation/nifti1fields/nifti1fields_pages/vox_offset.html).
Do you really need it here ? Sometimes vox_offset can be 352 and everything fall appart. If you don't need it please put it back at -0.5. We can also have a discussion with @GuillaumeTh  about this.

**Comment by AntoineTheb** [comment]
@arnaudbore I don't think I understand. If vox_offset can be 352, why always set it to -0.5 ? Shouldn't it be 352 ? Moreover, the offset should stay the same even if the dataset is downsampled or upsampled, no ? Also, when I tested it on the fibercup (which doesn't have an offset of -0.5) without keeping the offset, the downsampled volume ended up being, well, offset, to its original. I don't see why we would keep it at -0.5 always. We can discuss it, I guess I'm missing something.

**Comment by AntoineTheb** [comment]
@arnaudbore any update on this ? It will be needed very soon.

File: scilpy/image/resample_volume.py [hunk 5]
```diff
@@ -12,8 +12,8 @@ def _interp_code_to_order(interp_code):
...
+                    interp='lin', enforce_dimensions=False, offset=-0.5):
```
[the remark this answers is not in the document]
Comment (AntoineTheb) [reply scilpy/image/resample_volume.py:16]: Resolved in b3d1992

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Comment by scil-admin** [comment]
Build passed ! Good Job :beers: !

**Review by arnaudbore — APPROVED** [review]
This PR is really sensitive but thanks to @AntoineTheb great job's it seems pretty smooth. 
Let's move forward !

**Pull request merged**